File: XuacsInItsWay/Orbwalker.cpp
#include "Orbwalker.h"
#include "Globals.h"

#include "Mouse.h"

#include <iostream>


bool find_it_in_list(std::vector<CObject> list, int addr)
{
	for (int i = 0; i < list.size(); i++)
	{
		if (list[i].basePosition == addr)
			return true;
	}

	return false;

}

//
//int PredictHealth(int object)
//{
//	ReadEntities(false);
//
//	for (int i = 0; i < SObject::allMissiles.size(); i++)
//	{
//		if (SObject::allMissiles[i].basePosition <= 0) continue;
//
//		std::cout << "missile address: " << std::hex << SObject::allMissiles[i].basePosition << std::dec << std::endl;
//
//		//for (int j = 0; j < 100000; j+=4)
//		//{
//		//	int addr = Management::Manager.Read<int>(Management::Player_List[i].basePosition + j);
//
//		//	if (addr <= 0) continue;
//
//		//	if (find_it_in_list(Management::Player_List, addr))
//		//		std::cout << "offset is: " << std::hex << j << std::dec << ", address is: " << std::hex << addr << std::dec << std::endl;
//
//		//}
//
//	}
//
//	//for(int i = 0; i < 	Management::Player_List.size();i++)
//	//{
//	//	if (Management::Player_List[i].basePosition <= 0) continue;
//
//	//	std::cout << "hero address: " << std::hex << Management::Player_List[i].basePosition << std::dec << std::endl;
//	//	std::cout << "hero source_index: " << std::hex << Management::Player_List[i].source_index << std::dec << std::endl;
//	//	//std::cout << "local_player network id: " << std::hex << Management::local_player.network_id << std::dec << std::endl;
//
//	//	int active_spell = Management::Manager.Read<int>(Management::Player_List[i].spell_book + 0x20);
//
//	//	if (active_spell <= 0) continue;
//
//	//	std::cout << "active_spell_address: "<<std::hex << active_spell <<std::dec<< std::endl;
//	//	std::cout << "local_player_source_index: "<<std::hex << Management::local_player.source_index <<std::dec<< std::endl;
//
//
//	//	//for (int j = 0; j < 100000; j+=4)
//	//	//{
//	//	//	int addr = Management::Manager.Read<int>(Management::Player_List[i].basePosition + j);
//
//	//	//	if (addr <= 0) continue;
//
//	//	//	if (find_it_in_list(Management::Player_List, addr))
//	//	//		std::cout << "offset is: " << std::hex << j << std::dec << ", address is: " << std::hex << addr << std::dec << std::endl;
//
//	//	//}
//
//	//}
//
//
//	return 0;
//}




int waitBeforeAttack = 0, waitBeforeWalk = 0;

DWORD cacheTarget = 0;
bool one_time = true;

 void Orbwalker::LaneClear()
{
	 int cur_time = GetTickCount();

	if (cur_time > waitBeforeWalk)
	{
		CObject local_player = Management::local_player;

		if (Management::under_mouse_object.network_id != 0)
		{
			if (Management::under_mouse_object.team == local_player.team)
				Mouse::RightClick();
		}
		else
			Mouse::RightClick();
		
	
		if (cur_time > waitBeforeAttack)
		{

			int count_ally_minion = 0;
			std::vector<CObject> entitiesx = GetMobsInfield(local_player.attack_range + 100, local_player, local_player.team, &count_ally_minion);

			if (entitiesx.size() > 0)
			{
				std::vector<CObject> entities = GetOnlyAttackable(entitiesx);

				CObject target, target_close_to_die;
				target_close_to_die.network_id = 0;
				target.network_id = 0;

				if (target.network_id != local_player.network_id)
				{
					if (target.network_id == 0)
					{
						target = GetKillableInArray(entities, local_player.base_attack + local_player.bonus_attack, Physical, &target_close_to_die);

						if (target.network_id == 0)
						{
							if (target_close_to_die.network_id != 0 && count_ally_minion > 1)
							{
								//target = target_close_to_die;
								return;
							}
							else {
								target = GetHighestHealthInArray(entities);
							}

						}
					}


					if (target.network_id == 0)
						return;


					D3DXVECTOR2 screenPos = WorldToScreen(target.position);



					if (screenPos.x != 0 || screenPos.y != 0)
					{
						cacheTarget = target.network_id;

						//std::cout << "health: " << target.health << std::endl;
						//std::cout << "max_health: " << target.max_health << std::endl;
						//std::cout << "base_attack: " << target.base_attack << std::endl;

						Mouse::ScriptClick(screenPos.x, screenPos.y);
						/*						DWORD temp1 = Manager.Read<DWORD>(local_player.basePosition + 0x2538);
						int temp2 = Manager.Read<DWORD>(temp1 + 0x1C);

						int v12 = Manager.Read<float>(temp2 + 0x234);*/

						double attackCastDelay = AttackDelay(local_player.basePosition, local_player.attack_speed_mod);
						//std::cout << "attack cast delay: " << attackCastDelay << std::endl;

						double attackCastDelayWalk = AttackCastDelay(local_player.basePosition, 64);
						//std::cout << "attackCastDelayWalk: " << attackCastDelayWalk << std::endl;

						int ping = Management::Manager.Read<int>(Management::Manager.Read<int>(Management::Manager.Read<int>(Management::LolClient.dwBase + 0x22D8D0C) + 0x68) + 0x18);

						std::cout << "ping: " << ping << std::endl;

						waitBeforeWalk = cur_time + (int)(attackCastDelayWalk * 10.0f * Orbwalker::adjust_for_attack_cast_delay) + ping / 2; // wait for attack anim
						waitBeforeAttack = cur_time + (int)(attackCastDelay * 10.0f * Orbwalker::adjust_for_attack_delay) + ping / 2; // wait for attack time


						bool thing = false;
						std::cout << "time: " << waitBeforeWalk - cur_time << std::endl;

						for (int try_chance = waitBeforeWalk - cur_time; try_chance > 0; try_chance--) {

							if (Management::Manager.Read<int>(Management::local_player.basePosition + 0x28C8) != 0)
							{
								thing = true;
								break;
							}

							Sleep(1);
						}


						if (!thing)
						{
							waitBeforeWalk = cur_time + 50;
							waitBeforeAttack = 0;
						}
					}
				}
			}
		}

	}
	//else
	//{
	//	int diff = waitBeforeWalk - GetTickCount();
	//	if (diff >= 0)
	//		Sleep(diff);

	//	if (Management::Manager.Read<int>(Management::local_player.basePosition + 0x28C8) == 0 && one_time)
	//	{


	//		waitBeforeWalk = 0;
	//		waitBeforeAttack = 0;

	//		one_time = false;
	//	}
	//}
}


void Orbwalker::Combo()
{
	int cur_time = GetTickCount();

	if (cur_time >= waitBeforeWalk)
	{
		one_time = true;
		CObject local_player = Management::local_player;

		if (Management::under_mouse_object.network_id != 0)
		{
			if (Management::under_mouse_object.team == local_player.team)
				Mouse::RightClick();
		}
		else
			Mouse::RightClick();
			

		if (cur_time >= waitBeforeAttack)
		{
			std::vector<CObject> entitiesx = GetHeroesInField(local_player.attack_range + 100, local_player, local_player.team);

			if (entitiesx.size() > 0)
			{
				std::vector<CObject> entities = GetOnlyAttackable(entitiesx);


				CObject target;

				std::vector<CObject> percent_entities = GetDueToHealthPercent(entities, local_player);

				if(percent_entities.size() > 0)
					target = GetTargetDueToPriority(percent_entities, local_player);
				else
					target = GetTargetDueToPriority(entities, local_player);
				

				if (target.network_id == 0)
					target = GetClosestOneInArray(entities, local_player);

				if (Management::current_target.network_id != 0)
				{
					float diff = (Management::current_target.position - local_player.position).sqrMagnitude();

					if (diff <= local_player.attack_range + 100)
					{
						target = Management::current_target;
					}
				}


				if (target.network_id != local_player.network_id && target.network_id != 0)
				{

					D3DXVECTOR2 screenPos = WorldToScreen(target.position);

					if (screenPos.x != 0 || screenPos.y != 0)
					{
						cacheTarget = target.network_id;

						//std::cout << "health: " << target.health << std::endl;
						//std::cout << "max_health: " << target.max_health << std::endl;
						//std::cout << "base_attack: " << target.base_attack << std::endl;
						//std::cout << "visible: " << target.visible << std::endl;

						Mouse::ScriptClick(screenPos.x, screenPos.y);
						double attackCastDelay = AttackDelay(local_player.basePosition, local_player.attack_speed_mod);

						double attackCastDelayWalk = AttackCastDelay(local_player.basePosition, 64);
						//std::cout << "attackCastDelayWalk: " << attackCastDelayWalk*1000 << std::endl;
						//std::cout << "attack cast delay: " << attackCastDelay*1000 << std::endl;

						int ping = Management::Manager.Read<int>(Management::Manager.Read<int>(Management::Manager.Read<int>(Management::LolClient.dwBase + 0x22D8D0C) + 0x68) + 0x18);

						waitBeforeWalk = cur_time + (int)(attackCastDelayWalk * 10.0f * Orbwalker::adjust_for_attack_cast_delay) + ping / 2; // wait for attack anim
						waitBeforeAttack = cur_time + (int)(attackCastDelay * 10.0f * Orbwalker::adjust_for_attack_delay) + ping / 2; // wait for attack time

						bool thing = false;

						std::cout << "time: " << waitBeforeWalk - cur_time << std::endl;

						for (int try_chance = waitBeforeWalk - cur_time; try_chance > 0; try_chance--) {

							if (Management::Manager.Read<int>(Management::local_player.basePosition + 0x28C8) != 0)
							{
								thing = true;
								break;
							}

							Sleep(1);
						}

						std::cout << "thing: " << thing << std::endl;


						if (!thing)
						{
							waitBeforeWalk = cur_time + 50;
							waitBeforeAttack = 0;
						}

						//Sleep(waitBeforeWalk - cur_time + 50);

						//PredictHealth(target.basePosition);

					}
				}
			}
		}

	}
	//else
	//{
	//	int diff = waitBeforeWalk - GetTickCount();
	//	if (diff >= 0)
	//		Sleep(diff);

	//	if (Management::Manager.Read<int>(Management::local_player.basePosition + 0x28C8) == 0 && one_time)
	//	{


	//		waitBeforeWalk = 0;
	//		waitBeforeAttack = 0;

	//		one_time = false;
	//	}
	//}
}

void Orbwalker::LastHit()
{
	int cur_time = GetTickCount();
	if (cur_time >= waitBeforeWalk)
	{
		CObject local_player = Management::local_player;

		if (Management::under_mouse_object.network_id != 0)
		{
			if (Management::under_mouse_object.team == local_player.team)
				Mouse::RightClick();
		}
		else
			Mouse::RightClick();


		if (cur_time >= waitBeforeAttack)
		{

			std::vector<CObject> entitiesx = GetMobsInfield(local_player.attack_range + 100, local_player, local_player.team);

			if (entitiesx.size() > 0)
			{
				std::vector<CObject> entities = GetOnlyAttackable(entitiesx);

				float damage = local_player.base_attack + local_player.bonus_attack;


				CObject target = GetKillableInArray(entities, damage, Physical);

				if (target.network_id != local_player.network_id && target.network_id != 0)
				{
					D3DXVECTOR2 screenPos = WorldToScreen(target.position);

					if (screenPos.x != 0 || screenPos.y != 0)
					{
						cacheTarget = target.network_id;


						Mouse::ScriptClick(screenPos.x, screenPos.y);
						double attackCastDelay = AttackDelay(local_player.basePosition, local_player.attack_speed_mod);
						//std::cout << "attack cast delay: " << attackCastDelay << std::endl;

						double attackCastDelayWalk = AttackCastDelay(local_player.basePosition, 64);
						//std::cout << "attackCastDelayWalk: " << attackCastDelayWalk << std::endl;

						int ping = Management::Manager.Read<int>(Management::Manager.Read<int>(Management::Manager.Read<int>(Management::LolClient.dwBase + 0x22D8D0C) + 0x68) + 0x18);

						waitBeforeWalk = cur_time + (int)(attackCastDelayWalk * 10.0f * Orbwalker::adjust_for_attack_cast_delay) + ping / 2; // wait for attack anim
						waitBeforeAttack = cur_time + (int)(attackCastDelay * 10.0f * Orbwalker::adjust_for_attack_delay) + ping / 2; // wait for attack time


						bool thing = false;

						for (int try_chance = waitBeforeWalk - cur_time; try_chance > 0; try_chance--) {

							if (Management::Manager.Read<int>(Management::local_player.basePosition + 0x28C8) != 0)
							{
								thing = true;
								break;
							}

							Sleep(1);
						}

						if (!thing)
						{
							waitBeforeWalk = cur_time + 50;
							waitBeforeAttack = 0;
						}
					}
				}
			}
		}
	}
	//else
	//{
	//	int diff = waitBeforeWalk - GetTickCount();
	//	if (diff >= 0)
	//		Sleep(diff);

	//	if (Management::Manager.Read<int>(Management::local_player.basePosition + 0x28C8) == 0 && one_time)
	//	{


	//		waitBeforeWalk = 0;
	//		waitBeforeAttack = 0;

	//		one_time = false;
	//	}
	//}
}


void Orbwalker::Update()
{
	if (GetAsyncKeyState(4) & 1)
	{
		CObject under_mouse_object = CObject::GetUnderMouseObject();
		if (isHero(under_mouse_object.basePosition))
		{
			if (under_mouse_object.team != CObject::GetLocalPlayer().team)
			{
				Management::current_target = under_mouse_object;
			}
			else
			{
				Management::current_target.network_id = 0;
				Management::current_target.basePosition = 0;
			}

		}
		else
		{
			Management::current_target.network_id = 0;
			Management::current_target.basePosition = 0;
		}

	}


	if (GetAsyncKeyState(harras_key) || GetAsyncKeyState(combo_key) || GetAsyncKeyState(last_hit_key))
	{

		Management::local_player = CObject::GetLocalPlayer();
		Management::under_mouse_object = CObject::GetUnderMouseObject();


		if (GetAsyncKeyState(combo_key))
		{
			ReadHeroes();

			Combo();
		}
		else if (GetAsyncKeyState(last_hit_key))
		{
			ReadEntities(false);

			LastHit();
		}
		else if (GetAsyncKeyState(harras_key))
		{
			ReadEntities(false);

			LaneClear();
		}

	}


	Management::wait_for_usage = false;
}


int Orbwalker::harras_key = 'V';
int Orbwalker::combo_key = ' ';
int Orbwalker::last_hit_key = 'X';
float Orbwalker::adjust_for_attack_cast_delay = 100;
float Orbwalker::adjust_for_attack_delay = 100;

File: XuacsInItsWay/orbwalker.lua

close_to_die = CObject()

close_to_die.network_id = 0

function Loop()

  if IsKeyPressed(string.byte(" ")) then
  
    read_my_hero()
    read_heroes()
    
    Combo()
  
  elseif IsKeyPressed(string.byte("V")) then
  
    read_my_hero()
    read_entities()
    
    LaneClear()
  
  elseif IsKeyPressed(string.byte("X")) then
  
    read_my_hero()
    read_entities()
    
    LastHit()
  
  end
end






attack_cast_delay = 0
attack_delay = 0

cur_time = 0

function Combo()
  
  cur_time = now()
  
  if cur_time >= attack_cast_delay then
  
    if underMouseObject.network_id ~= 0 then
      
      if underMouseObject.team == myHero.team or underMouseObject:IsTargetable() == false then
        
        RightClick()
        
      end
      
    else
      
      RightClick()
      
    end
  
    if cur_time >= attack_delay then
    
      local target = CObject()
      target.network_id = 0

      local infield_entities = GetEntitiesInField(all_champions, myHero.position, myHero.attack_range + 100)
    
      local attackable_entities = GetOnlyAttackable(infield_entities)
    
      local few_hp_entities = GetDueToHealthPercent(attackable_entities)
      
      if(few_hp_entities:size() > 0) then
        target = GetTargetDueToPriority(few_hp_entities)
      else
        target = GetTargetDueToPriority(attackable_entities);
      end
            
      if(target.network_id == 0) then
        
        target = GetClosestOneInArray(attackable_entities)
        
      end
    
      if target.network_id ~= 0 then
            
        local screen_pos = W2S(target.position)
      
        ScriptClick(screen_pos.x, screen_pos.y)
      
        attack_delay = cur_time + myHero:GetAttackDelay() * attack_delay_multiplier * 10 + GetPing() / 2
        attack_cast_delay = cur_time + myHero:GetAttackCastDelay() * attack_cast_delay_multiplier * 10 + GetPing() / 2
      
      
        local thing = false
        for i = attack_cast_delay - cur_time, 0, -1 do
          
          if myHero:IsAttacking() then
            
            thing = true
            break
            
          end
          
          sleep(1)
        end
        
        if thing == false then
          
          attack_cast_delay = cur_time + 25
          attack_delay = 0
          
        end
      
      end
    
    end
  end
end


function LaneClear()
  
  
  
  cur_time = now()
  
  if cur_time >= attack_cast_delay then
  
    if underMouseObject.network_id ~= 0 then
      
      if underMouseObject.team == myHero.team or underMouseObject:IsTargetable() == false then
        
        RightClick()
        
      end
      
    else
      
      RightClick()
      
    end
    
    
  
    if cur_time >= attack_delay then
    
      local target = CObject()
      target.network_id = 0

      local infield_entities = GetEntitiesInField(all_mobs, myHero.position, myHero.attack_range + 100)
    
      local attackable_entities = GetOnlyAttackable(infield_entities)


      target = GetKillableInArray(attackable_entities, myHero.base_attack + myHero.bonus_attack)
      
      if target.network_id == 0 and close_to_die.network_id ~= 0 then
        
      else

      
        

      
      if(target.network_id == 0) then
        
        target = GetHighestHealthInArray(attackable_entities);
        
      end
            
      if(target.network_id == 0) then
        
        target = GetClosestOneInArray(attackable_entities)
        
      end
    
      if target.network_id ~= 0 then
            
        local screen_pos = W2S(target.position)
      
        ScriptClick(screen_pos.x, screen_pos.y)
      
        
      
        attack_delay = cur_time + myHero:GetAttackDelay() * attack_delay_multiplier * 10 + GetPing() / 2
        attack_cast_delay = cur_time + myHero:GetAttackCastDelay() * attack_cast_delay_multiplier * 10 + GetPing() / 2
      
      
        local thing = false
        for i = (attack_cast_delay - cur_time), 0, -1 do
          
          if myHero:IsAttacking() then
            
            thing = true
            break
            
          end
          sleep(1)
        end
        
        if thing == false then
          
          attack_cast_delay = cur_time + 25
          attack_delay = 0
          
        end
      
      end
    end
    end
  end
end


function LastHit()
  
  cur_time = now()
  
  if cur_time >= attack_cast_delay then
  
    if underMouseObject.network_id ~= 0 then
      
      if underMouseObject.team == myHero.team or underMouseObject:IsTargetable() == false then
        
        RightClick()
        
      end
      
    else
      
      RightClick()
      
    end
    
    
  
    if cur_time >= attack_delay then
    
      local target = CObject()
      target.network_id = 0

      local infield_entities = GetEntitiesInField(all_mobs, myHero.position, myHero.attack_range + 100)
    
      local attackable_entities = GetOnlyAttackable(infield_entities)
    
      target = GetKillableInArray(attackable_entities, myHero.base_attack + myHero.bonus_attack)

      if target.network_id ~= 0 then
            
        local screen_pos = W2S(target.position)
      
        ScriptClick(screen_pos.x, screen_pos.y)
      
        attack_delay = cur_time + myHero:GetAttackDelay() * attack_delay_multiplier * 10 + GetPing() / 2
        attack_cast_delay = cur_time + myHero:GetAttackCastDelay() * attack_cast_delay_multiplier * 10 + GetPing() / 2
      
      
        local thing = false
        for i = attack_cast_delay - cur_time, 0, -1 do
          
          if myHero:IsAttacking() then
            
            thing = true
            break
            
          end
          
          sleep(1)
        end
        
        if thing == false then
          
          attack_cast_delay = cur_time + 50
          attack_delay = 0
          
        end
      
      end
    
    end
  end
end

function CalculateDmgToObj(obj, Damage) 
	local myObj = myHero

	local EffectiveHealth = obj.predicted_health
  
  if obj.predicted_after_time > now() then
    EffectiveHealth = obj.health
  end
  
	local resists = 0

	local ourDamage = Damage

	resists = obj.armor

	EffectiveHealth = EffectiveHealth * (1.0 + (resists / 100.0))

	return EffectiveHealth - ourDamage
end

function GetHighestHealthInArray(list)

	local mMax = -99999
  
  local entity = CObject()
	entity.network_id = 0

	for i = 0, list:size()-1, 1 do
	
		if (list:at(i).health > mMax) then
			mMax = list:at(i).health
			entity = list:at(i)
		end
  end
	

	return entity
end

function GetHighestMaxHealthInArray(list_cur)

	local mMax = -99999

	local entitya = CObject()
  entitya.network_id = 0

	for i = 0, list_cur:size()-1, 1 do
	
		if list_cur:at(i).max_health > mMax then
			mMax = list_cur:at(i).max_health
			entitya = list_cur:at(i)
		end
    
  end
	

	return entitya
end

function GetKillableInArray(list, damage)
  
  local len = list:size() - 1

  local killable_list_loc = CObjectList()
  killable_list_loc:clear()
  
  log("test-1")

  PredictHealth(list)
    log("test-2")
close_to_die = CObject()
  close_to_die.network_id = 0
    log("test-3")
    
  for i = 0, len, 1 do
    
    local entityloc = list:at(i)
    log("test-4")
    local last_health = CalculateDmgToObj(entityloc, damage)
    log("test-5")
    if( last_health <= 0 ) then
      killable_list_loc:push(entityloc)
    end
    log("test-6")
    if(last_health > 0 and last_health <= damage * 3) then
      close_to_die = entityloc
    end
    log("test-7")
    if close_to_die.network_id == 0 then
      
      if(last_health > 0 and last_health <= damage * 4 and entityloc.health ~= entityloc.predicted_health) then
        close_to_die = entityloc
      end
      
    end
    log("test-8")
  end
  log("test-9")
  for i = 0, killable_list_loc:size()-1 do
    
    if killable_list_loc:at(i).health ~= killable_list_loc:at(i).predicted_health then
      return killable_list_loc:at(i)
    end
    
  end
  log("test-10")
  return GetHighestMaxHealthInArray(killable_list_loc)

end



function GetEntitiesInField(list, position, radius)
  
  local new_list = CObjectList()
  
  local len = list:size() - 1

  for i = 0, len, 1 do
    
    local entity = list:at(i)
 
    local dist = DistanceBetween(position, entity.position)
  
    if dist <= radius and entity:IsValid() then
            
      new_list:push(entity)     
      
    end
    
  end
  
  
  return new_list

end


function GetOnlyAttackable(list)
  
  local new_list = CObjectList()
  
  local len = list:size() - 1

  for i = 0, len, 1 do
    
    local entity = list:at(i)
    
    if entity.team ~= myHero.team and entity.health > 0 and entity.max_health > 5 and entity:IsVisible() and entity:IsValid() then
            
      new_list:push(entity)     
      
    end
    
  end
  
  
  return new_list

end





function GetDueToHealthPercent(list)
  
  local new_list = CObjectList()
  
  local len = list:size() - 1

  for i = 0, len, 1 do
    
    local entity = list:at(i)
    
    local percent = (entity.health / entity.max_health) * 100;
    
    if percent <= attack_under_hp_percent then
            
      new_list:push(entity)     
      
    end

  end
  
  
  return new_list

end


function GetTargetDueToPriority(list)
  
  local min = 999
  local index = -1

  local len = list:size() - 1

  
  
  for i = 0, len, 1 do
    
    local entity = list:at(i)
    
    if priority_list:find( entity.network_id ) ~= priority_list:endv() then
      
      if priority_list:at( entity.network_id ) < min then
        
        min = priority_list:at( entity.network_id )
        index = i
        
      end
      
    end
    
  end
  
  if index ~= -1 then
    return list:at(index)
  end
  
  entity = CObject()
  entity.network_id = 0
  
  return entity

end



function GetClosestOneInArray(list)
  
  local glob_entity = CObject()
  glob_entity.network_id = 0
  local max = 99999999

  local len = list:size() - 1

  for i = 0, len, 1 do
    
    local entity = list:at(i)
    
    local dist = DistanceBetween(myHero.position, entity.position)
    
    if dist < max then
      
      max = dist
      glob_entity = entity
      
    end 
    
  end
  

  
  return glob_entity

end


function DistanceBetween(first, second)
  
  local diff = Vec3 ( ( second.x - first.x ), ( second.y - first.y ), ( second.z - first.z ) )
  
  local dist = diff.x * diff.x + diff.z * diff.z
  
  return math.sqrt(dist)
end


function find_in_vector(list, to_find)
  
  local len = list:size() - 1
  
  for i = 0, len do
    
    if list:at(i).source_index == to_find.source_index then
      return i
    end
    
  end
  
  
  return -1
  
end


function PredictHealth(to_edit_list)
  
  local list = all_missiles
  
    for i = 0, list:size() - 1 do
    
      local missile = list:at(i)
      
      caster = GetObjectByIndex(missile.source_index)
      target = GetObjectByIndex(missile.target_id)
      
      if caster.source_index >= 0 and caster.source_index <= 10000 and target.source_index >= 0 and target.source_index <= 10000 then
      
      index = find_in_vector(to_edit_list, target)
      
      if index ~= -1 then

        local damage = caster.base_attack + caster.bonus_attack

        to_edit_list:at(index).predicted_health = to_edit_list:at(index).health - damage
        to_edit_list:at(index).predicted_after_time = now() + DistanceBetween(missile.position, missile.end_pos) * 0.5
      end
      
      end
      
    end

    

end

